### PR TITLE
embed libzdnn in model.so

### PR DESCRIPTION
**Pull request by gongsu832**
- Build libzdnn.a with -fPIC and embed in model.so when -maccel=NNPA specified
- Add CompilerConfigMap to store states associated with certain options
- Move options in main() to CompilerOptions.cpp
- Fix compiler warning in Stickify.cpp

Signed-off-by: Gong Su <gong_su@hotmail.com>

### CODE REVIEW

**Comment by gongsu832** [comment]
@tungld @negiyas libzdnn.a is built directly from https://github.com/IBM/zDNN so there is no need for `third_party/zdnn-lib` anymore. But I kept it just in case and we can remove it if you confirm it's OK.

**Comment by tungld** [comment]
> there is no need for third_party/zdnn-lib anymore. But I kept it just in case and we can remove it if you confirm it's OK.

We still use `zdnn.h` from `third_party/zdnn-lib`, and it is only that file.

File: src/Accelerators/NNPA/CMakeLists.txt [hunk 1]
```diff
@@ -19,6 +19,9 @@ else()
   set(ZDNN_LIBRARY_ENABLED 0)
 endif()
 
+include(zdnn.cmake)
+setup_zdnn(v0.4.0)
```
Comment (tungld) [inline src/Accelerators/NNPA/CMakeLists.txt:23]: We have moved to v1.0.0

File: CMakeLists.txt [hunk 2]
```diff
@@ -36,16 +36,6 @@ if (MSVC)
...
 
-# Define preprocessor for accelerators.
-set(NNPA_ENABLED 0)
-if (ACCELERATORS_TO_BUILD)
-  foreach(t ${ACCELERATORS_TO_BUILD})
-    if (${t} STREQUAL "NNPA")
-      set(NNPA_ENABLED 1)
-    endif()
-  endforeach(t)
-endif(ACCELERATORS_TO_BUILD)
```
Comment (tungld) [inline CMakeLists.txt:47]: NNPA_ENABLED is used to trigger lit tests for NNPA in `test/mlir/accelerators/nnpa` folder. It's my mistake not updating the comment that is confusing.

Comment (gongsu832) [reply CMakeLists.txt:47]: Yes I just saw that. I'm trying to see if there is a better way to do this.

File: src/Accelerators/NNPA/CMakeLists.txt (hunk 1 above)
Comment (gongsu832) [reply src/Accelerators/NNPA/CMakeLists.txt:23]: OK. I will move to v1.0.0 and see if I can fix the zdnn.h problem.

File: src/Compiler/CompilerOptions.cpp [hunk 3]
```diff
@@ -150,6 +177,14 @@ llvm::cl::opt<bool> onnxOpTransformReport("onnx-op-transform-report",
     llvm::cl::desc("Report diagnostic info for op transform passes."),
     llvm::cl::init(false), llvm::cl::cat(OMPassOptions));
 
+// Configuration states associated with certain options.
+// For example, when maccel is specified, NNPA can register
+// dependent libdnn.
+// This is just a simple string to vector map currently.
+// If it gets more complicated in the future, it can be
+// replaced by a class of its own.
+std::map<std::string, std::vector<std::string>> CompilerConfigMap;
```
Comment (etiotto) [inline src/Compiler/CompilerOptions.cpp:186]: Wondering whether we could use LLVM containers here like SmallVector and StringRef.

File: src/Accelerators/NNPA/CMakeLists.txt [hunk 4]
```diff
@@ -44,4 +47,4 @@ add_onnx_mlir_library(NNPAAccel
   OMZHighOps
   OMZLowOps
   )
-
+add_dependencies(NNPAAccel zdnn)
```
Comment (sstamenova) [inline src/Accelerators/NNPA/CMakeLists.txt:50]: You can add DEPENDS in the `add_onnx_mlir_library` call instead.

Comment (gongsu832) [reply src/Accelerators/NNPA/CMakeLists.txt:50]: Perfect. Should have read MLIR.cmake more carefully. Thanks.

File: test/mlir/lit.site.cfg.py.in [hunk 5]
```diff
@@ -8,7 +8,7 @@ config.targets_to_build = "@TARGETS_TO_BUILD@"
 config.onnx_mlir_tools_dir = r"@ONNX_MLIR_TOOLS_DIR@"
 config.onnx_mlir_obj_root = r"@ONNX_MLIR_BIN_ROOT@"
 
-config.enable_nnpa= @NNPA_ENABLED@
+config.enable_nnpa= 0x0@NNPA_ENABLED@
```
Comment (tungld) [inline test/mlir/lit.site.cfg.py.in:11]: This is clever.

File: src/Accelerators/CMakeLists.txt [hunk 6]
```diff
@@ -8,6 +8,7 @@ if (ACCELERATORS_TO_BUILD)
     add_subdirectory(${t})
     list(APPEND ACCEL_LIST "${t}Accel")
   endforeach(t)
+  message(STATUS "  Accelerators              : ${ACCEL_LIST}")
```
Comment (sstamenova) [inline src/Accelerators/CMakeLists.txt:11]: Can you make this message(DEBUG) or message(VERBOSE)? The log is already pretty noisy.

Comment (gongsu832) [reply src/Accelerators/CMakeLists.txt:11]: The noise is mostly produced by third party packages like onnx, not onnx-mlir. I actually would like to see more about what was set and not set, like the summary produced by onnx. The useless lines are the `-- Performing Test ...` lines.

File: src/Compiler/CompilerOptions.cpp (hunk 3 above)
Comment (etiotto) [reply src/Compiler/CompilerOptions.cpp:159]: ok

**Comment by AlexandreEichenberger** [comment]
Are all of the issues resolved @sstamenova @etiotto @gongsu832 ?

**Comment by sstamenova** [comment]
> Are all of the issues resolved @sstamenova @etiotto @gongsu832 ?
> 
> In addition, the PR needs a manual merge to resolve a conflict.

I would still prefer to see the `cmake` changes that are unrelated to `libzdnn` in a separate PR. Right now looking at the title of the PR I would not expect it to have the changes it has.

**Review by tungld — APPROVED** [review]
LGTM.

**Pull request merged**